src/ArborX_DistributedTree.hpp

```diff
@@ -86,16 +87,24 @@ class DistributedTree
    *     - \c distances Computed distances (optional and only for nearest
    *       predicates).
    */
-  template <typename ExecutionSpace, typename Predicates, typename... Args>
-  void query(ExecutionSpace const &space, Predicates const &predicates,
+  template <typename ExecutionSpace, typename UserPredicates, typename... Args>
+  void query(ExecutionSpace const &space, UserPredicates const &user_predicates,
              Args &&...args) const
   {
-    static_assert(Kokkos::is_execution_space<ExecutionSpace>::value);
-    using Access = AccessTraits<Predicates, PredicatesTag>;
-    using Tag = typename Details::AccessTraitsHelper<Access>::tag;
+    static_assert(
+        KokkosExt::is_accessible_from<MemorySpace, ExecutionSpace>::value);
+
+    using Predicates = Details::AccessValues<UserPredicates, PredicatesTag>;
+    static_assert(
+        KokkosExt::is_accessible_from<typename Predicates::memory_space,
+                                      ExecutionSpace>::value,
+        "Predicates must be accessible from the execution space");
+
+    using Tag = typename Predicates::value_type::Tag;
     using DeviceType = Kokkos::Device<ExecutionSpace, MemorySpace>;
     Details::DistributedTreeImpl<DeviceType>::queryDispatch(
-        Tag{}, *this, space, predicates, std::forward<Args>(args)...);
+        Tag{}, *this, space, Predicates{user_predicates},
```

Comment:
Nitpick: just define predicates before dispatching.

Reply:
I don't understand. You want `Predicates predicates{user_predicates};` before `queryDispatch`?

Reply:
yes

Reply:
Do you want the same in BVH and BruteForce? It's the same code.

Reply:
Up to you. I mostly didn't like the formatting change here.